src/coreclr/jit/valuenum.cpp

```diff
@@ -8042,7 +8044,7 @@ void Compiler::fgValueNumberBlockAssignment(GenTree* tree)
             if (tree->OperIsInitBlkOp())
             {
                 ValueNum lclVarVN = ValueNumStore::NoVN;
-                if (isEntire && rhs->IsIntegralConst(0))
+                if (isEntire && rhs->IsIntegralConst(0) && !lhsVarDsc->IsHiddenBufferStructArg())
```

Comment:
What is the case that does not work right without this condition?

This feels like a workaround for a problem that exists elsewhere.

Reply:
If I recall correctly, several [libraries stress tests were failing](https://dev.azure.com/dnceng/public/_build/results?buildId=1610813&view=ms.vss-test-web.build-test-results-tab&runId=44869708&resultId=132273&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab) because we were assuming that the stack is zero initialized and propagate the zero value at the first use (after the call) ignoring that it is re-defined at the call. Since the value was zero at the first use (after the call), we would eliminate bunch of code making it functionally incorrect.  I will still double check if the problem persists after the changes I made after that point.

Reply:
> We were assuming a.Field.IntegerField is 0?

For the test that I pointed, yes.

Reply:
I haven't verified after I added https://github.com/dotnet/runtime/pull/64130/files#diff-5b83397bbbdd17bb9457998b520fdaaa474d165390985b66f32371561b6d0bacR3539 that this is needed or not. I will verify that.

Reply:
Interesting, the code in the PR would suggest to me it should be handled correctly. Will come back to you with what I find.

Reply:
I don't see any asmdiffs for windows/x64 at least after removing the 2 hidden buffer checks around zero-init.